<a href="https://colab.research.google.com/github/dennismathewjose/Financial-ChatBot--BCG-Gen-AI-Forage-/blob/main/BCG_GenAI_Financial_Analysis_10_K_reports.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

### Your task is to manually extract key financial data for the last three fiscal years from the 10-K filings of Microsoft, Tesla, and Apple. Following the data collection, you will use Python to analyze this data, focusing on trends and insights that could inform the development of an AI-powered financial chatbot.



# Loading and Pre Processing

In [1]:
import pandas as pd
import numpy as np

In [50]:
df = pd.read_csv('10-K_financial_figures.csv')
df.head()

Company  Year Total Revenue Net Income Total Assets Total Liabilities  \
0   Apple  2022       198,270     72,738      364,840           198,298   
1   Apple  2023       211,915     72,361      411,976           205,753   
2   Apple  2024       245,122     88,136      512,163           243,686   
3   Tesla  2022       394,328     99,803      352,755           302,083   
4   Tesla  2023       383,285     96,995     352,583            290,437   

  Cash Flow  
0    89,035  
1    87,582  
2   118,548  
3   122,151  
4   110,543

In [51]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9 entries, 0 to 8
Data columns (total 7 columns):
 #   Column             Non-Null Count  Dtype 
---  ------             --------------  ----- 
 0   Company            9 non-null      object
 1   Year               9 non-null      int64 
 2   Total Revenue      9 non-null      object
 3   Net Income         9 non-null      object
 4   Total Assets       9 non-null      object
 5   Total Liabilities  9 non-null      object
 6   Cash Flow          9 non-null      object
dtypes: int64(1), object(6)
memory usage: 636.0+ bytes


In [52]:
columns = ['Total Revenue','Net Income','Total Assets','Total Liabilities','Cash Flow']
#converting these columns to integer type
for col in columns:
    df[col] = df[col].str.replace(',','').astype(int)

# Trend and Growth Analysis

In [53]:
df['Revenue Growth (%)'] = df.groupby(['Company'])['Total Revenue'].pct_change() * 100

In [54]:
df['Net Grwoth (%)'] = df.groupby(['Company'])['Net Income'].pct_change() * 100

In [55]:
df

Company  Year  Total Revenue  Net Income  Total Assets  \
0      Apple  2022         198270       72738        364840   
1      Apple  2023         211915       72361        411976   
2      Apple  2024         245122       88136        512163   
3      Tesla  2022         394328       99803        352755   
4      Tesla  2023         383285       96995        352583   
5      Tesla  2024         391035       93736        364980   
6  Microsoft  2022          81462       12587         82338   
7  Microsoft  2023          96773       14974        106618   
8  Microsoft  2024          97690        7153        122070   

   Total Liabilities  Cash Flow  Revenue Growth (%)  Net Grwoth (%)  
0             198298      89035                 NaN             NaN  
1             205753      87582            6.882030       -0.518299  
2             243686     118548           15.669962       21.800417  
3             302083     122151                 NaN             NaN  
4             290437     110543           -2.800461       -2.813543  
5             308030     118254            2.021994       -3.359967  
6              36440      14724                 NaN             NaN  
7              43009      13256           18.795267       18.964010  
8              48390      14923            0.947578      -52.230533

In [56]:
df['Average Revenvue'] = df.groupby(['Company'])['Total Revenue'].transform('mean')

In [57]:
df

Company  Year  Total Revenue  Net Income  Total Assets  \
0      Apple  2022         198270       72738        364840   
1      Apple  2023         211915       72361        411976   
2      Apple  2024         245122       88136        512163   
3      Tesla  2022         394328       99803        352755   
4      Tesla  2023         383285       96995        352583   
5      Tesla  2024         391035       93736        364980   
6  Microsoft  2022          81462       12587         82338   
7  Microsoft  2023          96773       14974        106618   
8  Microsoft  2024          97690        7153        122070   

   Total Liabilities  Cash Flow  Revenue Growth (%)  Net Grwoth (%)  \
0             198298      89035                 NaN             NaN   
1             205753      87582            6.882030       -0.518299   
2             243686     118548           15.669962       21.800417   
3             302083     122151                 NaN             NaN   
4             290437     110543           -2.800461       -2.813543   
5             308030     118254            2.021994       -3.359967   
6              36440      14724                 NaN             NaN   
7              43009      13256           18.795267       18.964010   
8              48390      14923            0.947578      -52.230533   

   Average Revenvue  
0     218435.666667  
1     218435.666667  
2     218435.666667  
3     389549.333333  
4     389549.333333  
5     389549.333333  
6      91975.000000  
7      91975.000000  
8      91975.000000

In [97]:
df.fillna(0,inplace=True)

In [100]:
summary = df.groupby('Company').agg({
    'Revenue Growth (%)': 'mean',
    'Net Grwoth (%)': 'mean'
}).reset_index()

print("\nYear-over-Year Average Growth Rates (%):")
summary


Year-over-Year Average Growth Rates (%):


Company  Revenue Growth (%)  Net Grwoth (%)
0      Apple            7.517331        7.094040
1  Microsoft            6.580948      -11.088841
2      Tesla           -0.259489       -2.057837

## Financial Insights ( 2022 -  2024 )
### **Apple**
#### Revenue Growth:

- Moderate growth in 2023: **+6.88%**

- Stronger performance in 2024: **+15.67%**

#### Net Income:

Declined slightly in 2023 (**–0.52%**), followed by a strong rebound in 2024 (**+21.80%**).

#### Observation:

- Apple demonstrates resilient profitability and strong asset expansion, with Total Assets rising from **364M to $512.2M** over three years.

- Cash flow from operating activities peaked in 2024 (**$118.5M**), supporting a healthy earnings quality.

### **Tesla**
#### Revenue Trend:

- Slight dip in 2023 (**–2.80%**), marginal recovery in 2024 (**+2.02%**), suggesting revenue stagnation.

#### Net Income:

- Consistently declining over the period: from **99.8M (2022) → $93.7M (2024)**, with **–3.36%** drop in 2024.

#### Observation:

- Despite a flat asset base, Tesla maintains strong operating cash flow **(~$118M in 2024)**, signaling operational efficiency.

- However, declining net income may raise concerns about long-term profitability.

### **Microsoft**
#### Revenue Growth:

- Excellent growth in 2023 **(+18.79%)**, but almost stagnant in 2024 **(+0.95%)**.

#### Net Income:

- Sharp increase in 2023 **(+18.96%)**, followed by a significant drop in 2024 **(–52.23%)**.

#### Observation:

- Although Microsoft’s assets grew steadily, the 2024 income drop is alarming, possibly hinting at increased expenses, write-downs, or one-off charges.

- Operating cash flow is relatively low and flat **(~$14M)** compared to peers, which could limit reinvestment capacity.

# Profitability Analysis

In [95]:
df['Net Profit Margin'] = df['Net Income'] / df['Total Revenue'] * 100
import plotly.express as px
#creating a bar chart to see the net profit margin of each company during each year using a bar chart using plotly.express with colors of the bar as blue gradients
fig = px.bar(df, x='Year', y='Net Profit Margin', color='Company', barmode='group',title='Net Profit Margin of each company over the year 2022-2024')
fig.show()

Apple’s margins are again the strongest (~36% in 2022 and 2024), signaling premium pricing power and operational efficiency.

Tesla’s margins remain steady around 25%, a solid level given its manufacturing-heavy business.

Microsoft’s margins declined in 2024 (to ~7%), a sharp drop from earlier years, hinting at a significant dip in profitability.

In [96]:
#operating cash flow margins
df['Cash Flow Margin'] = df['Cash Flow'] / df['Total Revenue'] * 100
fig = px.bar(df, x = 'Year', y = 'Cash Flow Margin', color = 'Company', barmode = 'group',title = 'Cash flow margin of each company over the year 2022-2024')
fig.show()

Apple consistently leads with the highest cash flow margin, peaking in 2024 (nearly 48%).

Tesla remains stable around 30%, indicating reliable operating cash generation.

Microsoft has the lowest cash flow margins (~15% in 2024), which may reflect higher operational costs or less efficient cash conversion.

# Balance Sheet Ratios

In [92]:
from plotly.subplots import make_subplots
import plotly.io as pio
import plotly.graph_objects as go

custom_palette = ["#636EFA", "#3366cc","#1F77B4"]
pio.templates["custom_template"] = go.layout.Template(
    layout_colorway=custom_palette
)

pio.templates.default = "custom_template"

df['Debt to asset ratio'] = df['Total Liabilities'] / df['Total Assets']
fig = make_subplots(rows = 1, cols = 3, specs=[[{'type':'domain'},{'type':'domain'},{'type':'domain'}]])

fig.add_trace(
    go.Pie(labels = df[df['Company'] == 'Apple']['Year'], values = df[df['Company'] == 'Apple']['Debt to asset ratio'],name='Apple', title='Apple'),
    row = 1, col = 1
)
fig.add_trace(
    go.Pie(labels = df[df['Company'] == 'Tesla']['Year'], values = df[df['Company'] == 'Tesla']['Debt to asset ratio'],name = 'Tesla',title='Tesla'),
    row = 1, col = 2
)
fig.add_trace(
    go.Pie(labels = df[df['Company'] == 'Microsoft']['Year'], values = df[df['Company'] == 'Microsoft']['Debt to asset ratio'],name = 'Microsoft',title='Mircrosoft'),
    row = 1, col = 3
)
fig.update_layout(title_text="Debt to Asset Ratio of each Company from 2022 - 2024")
fig.show()

All three companies maintain debt levels under 40% of their total assets, indicating conservative use of leverage.

Apple shows a slight upward trend (from 32.9% to 35.8%), suggesting increased liabilities relative to asset growth.

Tesla has the most stable ratio, hovering near 33% across all years.

Microsoft’s debt ratio slightly decreased in 2023 and 2024, showing improved balance sheet strength.

| Metric               | Apple                            | Tesla                              | Microsoft                                           |
| -------------------- | -------------------------------- | ---------------------------------- | --------------------------------------------------- |
| **Debt Management**  | Slight increase, but healthy     | Very stable and low                | Gradual improvement                                 |
| **Cash Flow Margin** | Highest (peaked in 2024)         | Consistent (\~30%)                 | Weakest (\~15%)                                     |
| **Profitability**    | Strong and consistent            | Flat profit margins                | Sharp margin decline in 2024                        |
| **Overall Outlook**  | Financially robust and efficient | Operationally sound but plateauing | Revenue growth not translating into profits or cash |
